hl/src/H5LT.c

```diff
@@ -355,15 +355,28 @@ image_free(void *ptr, H5FD_file_image_op_t file_image_op, void *_udata)
 
             udata->vfd_ref_count--;
 
-            /* release the shared buffer only if indicated by the respective flag and there are no outstanding
-             * references */
-            if (udata->fapl_ref_count == 0 && udata->vfd_ref_count == 0 &&
-                !(udata->flags & H5LT_FILE_IMAGE_DONT_RELEASE)) {
-                free(udata->vfd_image_ptr);
-                udata->app_image_ptr  = NULL;
-                udata->fapl_image_ptr = NULL;
-                udata->vfd_image_ptr  = NULL;
-            } /* end if */
+            if (udata->fapl_ref_count == 0 && udata->vfd_ref_count == 0) {
```

Comment:
The default `udata_free` callback supplied by H5LT.c wraps around this and always decrements `udata->ref_count` and then only frees the structure when it drops to 0. Changing this code to always call `udata_free` here appears to pass the file image tests, though those aren't too extensive; I'm not sure yet if there are cases where that would be problematic. It seems one could unknowingly create a different memory leak with `H5Pget_file_image_callbacks` since it calls `udata_copy` without a chance for an `image_free` call somewhere, but this shouldn't happen with the default callbacks at least.

In my current understanding of the code, I don't believe there should be a case where `udata->ref_count == 0` and `udata->fapl_ref_count != 0 || udata->vfd_ref_count != 0`. The udata structure should always be owned by either a FAPL or the VFD, so we should always at least have `udata->ref_count == 1 && udata->fapl_ref_count == 1` OR `udata->ref_count == 1 && udata->vfd_ref_count == 1`.

Reply:
Ok I guess my suggestion would be to always call udata_free, after the `if (udata->fapl_ref_count == 0 && udata->vfd_ref_count == 0) {` block, and assert in udata_free that all the ref counts are 0 if udata->ref_count is 0